Issue 2218: Write customMetadata on ledgers created by DistributedLog

dmercuriali opened the pull request:
### Changes

Enable DistributedLog to attach custom metadata to underlying ledgers when writing logs.

The user can pass a LedgerMetadata object to DistributedLogManager#openLogWriter and DistributedLogManager#openAsyncLogWriter. The LedgerMetadata object is then used by the Allocator backing the BKLogWriteHandler every time it must allocate a new ledger.

Master Issue: #2218

> - [X] [skip bookkeeper-server bookie tests]: skip testing `org.apache.bookkeeper.bookie` in bookkeeper-server module.
> - [X] [skip bookkeeper-server client tests]: skip testing `org.apache.bookkeeper.client` in bookkeeper-server module.
> - [X] [skip bookkeeper-server replication tests]: skip testing `org.apache.bookkeeper.replication` in bookkeeper-server module.
> - [X] [skip bookkeeper-server tls tests]: skip testing `org.apache.bookkeeper.tls` in bookkeeper-server module.
> - [X] [skip bookkeeper-server remaining tests]: skip testing all other tests in bookkeeper-server module.
> - [ ] [skip integration tests]: skip docker based integration tests. if you make java code changes, you shouldn't skip integration tests.
> - [ ] [skip build java8]: skip build on java8. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> - [ ] [skip build java11]: skip build on java11. *ONLY* skip this when *ONLY* changing files under documentation under `site`.

eolivelli commented:
@dmercuriali palese writer in the description how the client application can set custom ledger metadata.
Otherwise from the PR it is hard to understand how this change can be used by users

Thanks for working on this

eolivelli commented on stream/distributedlog/core/src/main/java/org/apache/distributedlog/api/DistributedLogManager.java:
```diff
@@ -104,6 +105,17 @@
+     * @return result represents the open result
+     */
+    default CompletableFuture<AsyncLogWriter> openAsyncLogWriter(LedgerMetadata ledgerMetadata) {
+        throw new UnsupportedOperationException();
```
Return a failed future

eolivelli commented:
@dmercuriali the approach is good.
You have some javac error:
```

2020-01-30T17:03:44.4179776Z Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/httpcomponents/httpclient/4.4.1/httpclient-4.4.1.jar (721 kB at 6.3 MB/s)
2020-01-30T17:04:01.7976679Z [ERROR] COMPILATION ERROR : 
2020-01-30T17:04:01.7986913Z [ERROR] /home/runner/work/bookkeeper/bookkeeper/stream/distributedlog/core/src/main/java/org/apache/distributedlog/api/DistributedLogManager.java:[116,33] cannot find symbol
2020-01-30T17:04:01.7991275Z   symbol:   method failedFuture(java.lang.UnsupportedOperationException)
2020-01-30T17:04:01.7994806Z   location: class java.util.concurrent.CompletableFuture
2020-01-30T17:04:01.8023930Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project distributedlog-core: Compilation failure
2020-01-30T17:04:01.8033444Z [ERROR] /home/runner/work/bookkeeper/bookkeeper/stream/distributedlog/core/src/main/java/org/apache/distributedlog/api/DistributedLogManager.java:[116,33] cannot find symbol
2020-01-30T17:04:01.8039627Z [ERROR]   symbol:   method failedFuture(java.lang.UnsupportedOperationException)
2020-01-30T17:04:01.8044182Z [ERROR]   location: class java.util.concurrent.CompletableFuture
2020-01-30T17:04:01.8045720Z [ERROR] 
```

dmercuriali commented:
> @dmercuriali the approach is good.
> You have some javac error:
> 
> ```
> 
> 2020-01-30T17:03:44.4179776Z Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/httpcomponents/httpclient/4.4.1/httpclient-4.4.1.jar (721 kB at 6.3 MB/s)
> 2020-01-30T17:04:01.7976679Z [ERROR] COMPILATION ERROR : 
> 2020-01-30T17:04:01.7986913Z [ERROR] /home/runner/work/bookkeeper/bookkeeper/stream/distributedlog/core/src/main/java/org/apache/distributedlog/api/DistributedLogManager.java:[116,33] cannot find symbol
> 2020-01-30T17:04:01.7991275Z   symbol:   method failedFuture(java.lang.UnsupportedOperationException)
> 2020-01-30T17:04:01.7994806Z   location: class java.util.concurrent.CompletableFuture
> 2020-01-30T17:04:01.8023930Z [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project distributedlog-core: Compilation failure
> 2020-01-30T17:04:01.8033444Z [ERROR] /home/runner/work/bookkeeper/bookkeeper/stream/distributedlog/core/src/main/java/org/apache/distributedlog/api/DistributedLogManager.java:[116,33] cannot find symbol
> 2020-01-30T17:04:01.8039627Z [ERROR]   symbol:   method failedFuture(java.lang.UnsupportedOperationException)
> 2020-01-30T17:04:01.8044182Z [ERROR]   location: class java.util.concurrent.CompletableFuture
> 2020-01-30T17:04:01.8045720Z [ERROR] 
> ```

Fixed, my bad for building with jdk-11 locally.
Also added some tests.

eolivelli commented:
@dmercuriali I left one little comment, we have FutureUtils utility in order to create a failed Future

eolivelli commented on stream/distributedlog/core/src/main/java/org/apache/distributedlog/bk/LedgerMetadata.java:
```diff
@@ -0,0 +1,71 @@
+            for (Map.Entry<String, String> e : customMetadata.entrySet()) {
+                String value = e.getValue();
+                meta.put(e.getKey(), value == null
+                        ? null
```
It doesn't make sense to have a null value here.
We should fail with a null key or value

dmercuriali replied:
added a guard clause in LedgerMetadata#addCustomMetadata like the one for the key

eolivelli commented:
Some ci job failed PTAL

dmercuriali commented:
> [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.21.0:test (default-test) on project smoke: There was a timeout or other error in the fork -> [Help 1]

Can someone trigger a new run for the failing job?

eolivelli commented:
@dmercuriali can you please rebase onto latest master ?
I have fixed the integration tests issue (#2265 )

eolivelli commented:
Actually we have to merge this patch as well in order to have a fully working github actions env
https://github.com/apache/bookkeeper/pull/2266

eolivelli commented:
merged to master branch, it will be available from 4.11.0 onwards

thank you @dmercuriali
